python/cudf/cudf/_lib/interop.pyx

```diff
@@ -70,7 +70,7 @@ def to_dlpack(list source_columns):
     )
 
 
-cdef void dlmanaged_tensor_pycapsule_deleter(object pycap_obj):
+cdef void dlmanaged_tensor_pycapsule_deleter(object pycap_obj) noexcept:
```

Comment:
I agree this fixes the issue, but is it a lie? On line 80, we call `pycapsule.PyCapsule_GetPointer(pycap_obj, "dltensor")`. This _feels_ like it could raise an exception, certainly I am guessing that the try/except around getting `used_dltensor` is because that attribute might not exist.

If I were writing this in C, I would probably do something like:
```c

dlpack_tensor = PyCapsule_GetPointer(obj, "used_dltensor");
if (dlpack_tensor) {
    // Already used, we're done
    return;
else {
    // name didn't exist, let's pretend nothing happened
    PyErr_Clear();
    dlpack_tensor = PyCapsule_GetPointer(obj, "dltensor");
    if (dlpack_tensor) {
        // we need to call the deleter
        dlpack_tensor.deleter(dlpack_tensor)
    else {
        // something failed, let's pretend nothing happened
        PyErr_Clear();
    }
}
```

So in cython, something like:

```cython
cdef void deleter(object obj) noexcept:
    cdef DLManagedTensor *tensor = <DLManagedTensor*>0 # aside this is an unsafe cast since that might not be the rep of null, should use NULL
    try:
        tensor = <DLManagedTensor*>pycapsule.PyCapsule_GetPointer(obj, 'used_dltensor')
    except Exception:
        try:
            tensor = <DLManagedTensor*>pycapsule.PyCapsule_GetPointer(obj, 'dltensor')
            tensor.deleter(tensor)
        except Exception:
            pass
```

Reply:
You're right that the call to `PyCapsule_GetPointer` can _technically_ raise an exception. And because the destructor is [not expected to raise](https://github.com/cython/cython/blob/master/Cython/Includes/cpython/pycapsule.pxd#L16), we cannot propagate any exceptions from it. That is what the `noexcept` is doing for us: it is ensuring that Cython will not wrap calls to `deleter` with code that will check for and raise Python exceptions.

I think that adding the `try-except-pass` would do more harm than good. To see why, notice that neither of these Cython programs will raise a Python error, and both will print "hello":

```cython                                                                                                                                            
cdef void foo() noexcept:                                                                                                                                                         
    raise ZeroDivisionError()                                                                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                  
foo() 
```

```cython
cdef void foo() noexcept:                                                                                                                                                             try:                                                                                                                                                                          
        raise ZeroDivisionError()                                                                                                                                                 
    except:                                                                                                                                                                               pass                                                                                                                                                                      
                                                                                                                                                                    
foo()                                                                                                                                                                             
print("hello")
```

But, the first will print a warning about an exception being ignored. The second will be totally silent about any exceptions.

Reply:
This being said, should we expect the destructor to be able to propagate Python exceptions? I'm not sure when the destructor is expected to run in relation to the state of the Python runtime and if that would make sense.

Reply:
OK, makes sense.